File: src/main/java/com/matejdro/wearutils/preferences/AutoSummaryEditTextPreference.java
package com.matejdro.wearutils.preferences;

import android.content.Context;
import android.preference.EditTextPreference;
import android.preference.Preference;
import android.util.AttributeSet;

import com.matejdro.wearutils.miscutils.HtmlCompat;

public class AutoSummaryEditTextPreference extends EditTextPreference implements Preference.OnPreferenceChangeListener
{
    private String summaryFormat;

    public AutoSummaryEditTextPreference(Context context, AttributeSet attrs, int defStyleAttr)
    {
        super(context, attrs, defStyleAttr);
        init();
    }

    public AutoSummaryEditTextPreference(Context context, AttributeSet attrs)
    {
        super(context, attrs);
        init();
    }

    public AutoSummaryEditTextPreference(Context context)
    {
        super(context);
        init();
    }

    private void init()
    {
        summaryFormat = getSummary().toString();

        setOnPreferenceChangeListener(this);
    }

    @Override
    protected void onSetInitialValue(boolean restoreValue, Object defaultValue)
    {
        super.onSetInitialValue(restoreValue, defaultValue);
        onPreferenceChange(this, getText());
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue)
    {
        String summary = String.format(summaryFormat, (String) newValue);
        setSummary(HtmlCompat.fromHtml(summary));
        return true;
    }
}

File: src/main/java/com/matejdro/wearutils/preferences/compat/AutoSummaryEditTextPreference.kt
package com.matejdro.wearutils.preferences.compat

import android.content.Context
import android.os.Bundle
import android.util.AttributeSet
import androidx.preference.EditTextPreference
import androidx.preference.EditTextPreferenceDialogFragmentCompat
import androidx.preference.PreferenceDialogFragmentCompat
import com.matejdro.wearutils.miscutils.HtmlCompat

/**
 * Variant of [EditTextPreference] that automatically updates the preference's summary when
 * edited
 */
class AutoSummaryEditTextPreference(context: Context, attrs: AttributeSet) : EditTextPreference(context, attrs), PreferenceWithDialog {
    private val summaryFormat: String = summary.toString()

    init {
        val summary = String.format(summaryFormat, text ?: "")
        setSummary(HtmlCompat.fromHtml(summary))
    }

    override fun setText(text: String?) {
        super.setText(text)

        val summary = String.format(summaryFormat, getText() ?: "")
        setSummary(HtmlCompat.fromHtml(summary))
    }

    override fun createDialog(key: String): PreferenceDialogFragmentCompat {
        return AutoSummaryEditTextPreferenceDialog.create(key)
    }

    class AutoSummaryEditTextPreferenceDialog : EditTextPreferenceDialogFragmentCompat() {
        companion object {
            fun create(key: String): AutoSummaryEditTextPreferenceDialog {
                val fragment = AutoSummaryEditTextPreferenceDialog()

                val arguments = Bundle(1)
                arguments.putString(PreferenceDialogFragmentCompat.ARG_KEY, key)

                fragment.arguments = arguments
                return fragment
            }
        }
    }
}
